libraries/chain/db_market.cpp

```diff
@@ -585,11 +646,70 @@ asset database::match( const call_order_object& call,
 
+   auto maint_time = get_dynamic_global_properties().next_maintenance_time;
+   bool before_core_hardfork_342 = ( maint_time <= HARDFORK_CORE_342_TIME ); // better rounding
+
    auto settle_for_sale = std::min(settle.balance, max_settlement);
    auto call_debt = call.get_debt();
 
    asset call_receives   = std::min(settle_for_sale, call_debt);
-   asset call_pays       = call_receives * match_price;
+   asset call_pays       = call_receives * match_price; // round down here, in favor of call order, for first check
+                                                        // TODO possible optimization: check need to round up or down first
+
+   // Be here, the call order may be paying nothing.
+   bool cull_settle_order = false; // whether need to cancel dust settle order
+   if( call_pays.amount == 0 )
+   {
+      if( maint_time > HARDFORK_CORE_184_TIME )
+      {
+         if( call_receives == call_debt ) // the call order is smaller than or equal to the settle order
+         {
+            wlog( "Something for nothing issue (#184, variant C-1) handled at block #${block}", ("block",head_block_num()) );
+            call_pays.amount = 1;
+         }
+         else
+         {
+            if( call_receives == settle.balance ) // the settle order is smaller
+            {
+               wlog( "Something for nothing issue (#184, variant C-2) handled at block #${block}", ("block",head_block_num()) );
+               cancel_settle_order( settle );
+            }
+            else // neither order will be completely filled, perhaps due to max_settlement too small
+               wlog( "Something for nothing issue (#184, variant C-3) handled at block #${block}", ("block",head_block_num()) );
+            return asset( 0, settle.balance.asset_id );
+         }
+      }
+      else // TODO remove this warning after hard fork core-184
+         wlog( "Something for nothing issue (#184, variant C) occurred at block #${block}", ("block",head_block_num()) );
+   }
+   else // the call order is not paying nothing, but still possible it's paying more than minimum required due to rounding
+   {
+      if( !before_core_hardfork_342 )
+      {
+         if( call_receives == call_debt ) // the call order is smaller than or equal to the settle order
+         {
+            call_pays = call_receives ^ match_price; // round up here, in favor of settle order
+            // be here, we should have: call_pays <= call_collateral
+         }
+         else
+         {
+            // be here, call_pays has been rounded down
+
+            // be here, we should have: call_pays <= call_collateral
+
+            if( call_receives == settle.balance ) // the settle order will be completely filled, assuming we need to cull it
+               cull_settle_order = true;
+            // else do nothing, since we can't cull the settle order
+
+            call_receives = call_pays ^ match_price; // round up here to mitigate rounding issue (core-342)
```

Comment:
It is important to understand here that the newly rounded up call_receives cannot be greater than the old call_receives.

Reply:
Added your comment.

BTW I have a test case for this in the bsip38 PR, need to update atm due to the operator change though.